### DISCUSSION TITLE
Thoughts on creating a bounty for adding Vulkan support?

### DISCUSSION THREAD

**Discussion by sofakng** [andrei-drexler/ironwail#287] [General]
I'm just wondering if the developers had any thoughts on having a bounty for adding Vulkan support to Ironwail.

I completely understand why it wouldn't be a priority since OpenGL is working on Windows and Linux, but macOS has no method or translation layer to use this engine.  If Vulkan support was added then MoltenVK (Vulkan-to-Metal) could be used to allow this to work.

Let me know your thoughts :)

**Comment by mhQuake** [comment]
If you're going to add another API just for Macs, then I'd think just going straight to Metal support would surely be better.

**Reply by sofakng** [reply]
I don't think adding Metal would be worth it (although that would be great), but I can see adding Vulkan having a benefit other systems besides Macs.  Adding Vulkan would be the primary step and then MoltenVK would provide the macOS specific layer.

**Reply by drHyperion451** [reply]
At this point it isn't worth the effort. Yes, MoltenVK exists but it's not performant enough for low end computers. At this point, what's the point of porting a performance-orriented port if isn't performant?.

Metal porting is one possible option, but at this point it would be better to just be another fork, maybe ironwail-metal or something alike. I hope I could do it myself but I don't know how.

**Reply by ghost** [reply]
I have been a long time fan of vkQuake. The UI features in ironwail are great, but I kept thinking why does it not look as good as vkQuake??? See screenshots. The easiest comparison is if you first look at the shotgun. ironwail is dull and there is a big difference in the detail of the foregrip. Then you can see it in the rest of the scene. Nitpicky - Yes. I played various original and mod levels, playing the same level using both ports wondering what was wrong. Then I took the screenshots. Anyway, hats off to the source porters out there and the people making the fantastic mods. vkQuake is still the engine for me with ironwail being a close second. I wish vkQuake had the UI features that ironwail did and hopefully someone will advance a new fork.

ironwail
![ironwail](https://github.com/user-attachments/assets/489f799e-3513-43e2-862c-69629d43222f)

vkQuake
![vkQuake](https://github.com/user-attachments/assets/d4e97c32-8f1a-4138-9d3d-0043884acfe1)

**Reply by andrei-drexler** [reply]
Looking at the screenshots, the main difference is that vkQuake is using remastered models, while IW is using the original ones. The next IW version (coming soon) has support for remastered models as well - you can try a [more recent dev build](https://github.com/andrei-drexler/ironwail/actions/workflows/windows_ci.yml?query=branch%3Amaster+event%3Apush). Other than that, `gamma` and `contrast` settings are slightly different - if you match those, the rest of the scene should look identical.

**Reply by garoto** [reply]
`coming soon™`

**Comment by ghost** [comment]
HI Andrei - I wasn't expecting a response! So, thank you! I used the same id PAK file in both vkQuake and ironwail, the larger PAK0 rerelease. I did also try to match up the options as best I could, setting AA to 16x in both vs the 32x available in ironwail; brightness, contrast etc. Checked to make sure both ports were using my NVIDIA card, used the same monitor etc, etc...

I do have original id Quake PAKs, PAK0 & PAK1. I'll give those a try.

Thanks Again!
IT2138

On Oct 12, 2024, at 1:33 AM, Andrei Drexler ***@***.***> wrote:



Looking at the screenshots, the main difference is that vkQuake is using remastered models, while IW is using the original ones. The next IW version (coming soon) has support for remastered models as well - you can try a more recent dev build<https://github.com/andrei-drexler/ironwail/actions/workflows/windows_ci.yml?query=branch%3Amaster+event%3Apush>. Other than that, gamma and contrast settings are slightly different - if you match those, the rest of the scene should look identical.

**Comment by ghost** [comment]
Greetings Everyone! So I did a fresh install of vkQuake 1.31.2_win64 and ironwail 0.7.0-win64. I created two folders in each source port "id1-original" containing PAK0 and PAK1 and "id1-rerelease" with the single big PAK0 file. To get each screenshot I renamed the folder to just "id1" in sequence and ran each port. Each screenshot is on the same PC, same setup with the same monitor.

"id1" Original PAK files, PAK0, PAK1
Result: virtually the same between ironwail and vkQuake with ironwail being a bit duller. Again start with the foregrip on the shotgun.
ironwail
![ironwail with original PAK files PAK0, PAK1](https://github.com/user-attachments/assets/26beb130-c1f1-4867-8e5e-a6f6f640e3c6)

vkQuake
![vkquake with original PAK files PAK0, PAK1](https://github.com/user-attachments/assets/79907aec-b3ce-4c52-8f0b-3026535700d8)



"id1" Rerelease PAK file PAK0
Result: Difference as originally noted.
ironwail 
![ironwail with rerelease PAK file PAK0](https://github.com/user-attachments/assets/b11f1f9d-9716-4092-87f7-2bdf604c531c)

vkQuake
![vkquake with rerelease PAK file PAK0](https://github.com/user-attachments/assets/6b1c9d48-62c2-4e15-a408-5de158bdb9af)
-fd20-4ca4-80d5-910bf7872e23)

It appears that ironwail is limited or limiting the data of the file, whereas vkQuake is using everything. And I certainly do not mean to offend anyone, particularly Andrei. I'm just presenting what I'm seeing. Hope this helps!

**Comment by ghost** [comment]
Good Evening! With some more playing around I figured out what the difference is. vkQuake has an option in the Graphics section to utilize remastered models in the rerelease PAK file. The remasterd models are, of course, not available in the original PAK files which is why the shotgun looked the same in both ports. I could not find an option in ironwail or the ironwail console to turn on the remastered models of the rerelease PAK file. This could just be me. Anyway - more screenshots below to show the option. So, again my nit-picky perspective, visually I like vkQuake. As far as the UI? Its ironwail. I know ironwail can get some crazy frame rates. My system does great with either port!

Thanks Again!

vkQuake Remastered Model Setting
![vkQuake Remastered Model Setting](https://github.com/user-attachments/assets/ea3930ec-d4ef-4b8b-8aa2-20d35253a590)

vkQuake Rerelease PAK file with Remastered Graphic Model Setting
![vkQuake Rerelease PAK file with Remastered Graphic Model Setting](https://github.com/user-attachments/assets/cf8426d0-4355-4030-89ef-1216797c337a)


vkQuake Classic Model Setting
![vkQuake Classic Model Setting](https://github.com/user-attachments/assets/1e0d6864-33b9-452a-9558-a91360a8c84e)

vkQuake Rerelease PAK file with Classic Graphic Model Setting
![vkQuake Rerelease PAK file with Classic Graphic Model Setting](https://github.com/user-attachments/assets/284887e9-3382-4f5a-b88d-8d36e4c8b546)

**Reply by drHyperion451** [reply]
There is already the remaster models option inside ironwail. You just need to download or compile a dev build. I believe it will be out in 0.8.0. 
![imagen](https://github.com/user-attachments/assets/3d90b236-196b-4551-8e8f-bf5a858f4b5c)

**Comment by ghost** [comment]
Awesome - I'll give it a shot!

On Oct 13, 2024, at 5:24 AM, drHyperion451 ***@***.***> wrote:



There is already the remaster models option inside ironwail. You just need to download or compile a dev build. I believe it will be out in 0.8.0.
imagen.png (view on web)<https://github.com/user-attachments/assets/3d90b236-196b-4551-8e8f-bf5a858f4b5c>

**Comment by ghost** [comment]
Greetings! I was anxiously awaiting the release of IronWail .8 and I am loving the addition of the remastered models! Below is side by side comparison. vkQuake on the left and IronWail on the right (I also attached the original png files). I am a proponent of Vulkan but it is in my nature to go with non-standard/less popular things, of course there's always exceptions ;) I am certainly not a programmer at this level so it is with my sincerest gratitude to the IronWail team for their hard work!

![image](https://github.com/user-attachments/assets/3a49da28-7cfa-479d-8a36-6dee7912d479)

![vkquake0000](https://github.com/user-attachments/assets/a874f1b6-c174-491f-ac90-5897674420f5)

![start_2024-11-12_16-51-53](https://github.com/user-attachments/assets/4d2292f8-a7ba-494c-9dad-2467c8bde652)

**Comment by ghost** [comment]
Greetings Again - I know, I know, the comparison to vkQuake is getting old. BUT just one more time. Here is a side by side, vkQuake on the left and IronWail on the right. I adjusted the brightness and contrast ever so slightly, going 2 points up on brightness and 1 point up on contrast so they are 3 and 3 for the settings in IronWail. And they're just about identical except for difference in light in the background from the clouds. Anyway - just fascinating and awesome to me! I promise no more (at least until the next release ;)

![RetroJam7_VkQuake(left)_IronWail8(right)](https://github.com/user-attachments/assets/749cc845-a66b-4450-838f-9f680736e730)

**Comment by ghost** [comment]
Forgot to mention the game is Retro Jam 7, map: Confirmation of the Underworld.

**Comment by ghost** [comment]
Greetings - I guess I lied in my last post. I noticed a difference in the start map of Arcane Dimensions that I thought interesting. vkQuake is on the left and IronWail is on the right. Both using the same version of Arcane Dimensions BUT if you notice in the back at the portals there are light flakes and a bunch of smoke that is pouring out in the vkQuake version. IronWail goes through a smaller cycle of light flakes but no smoke that I have seen.

![Arcane Dimensions vkQuake vs IronWail](https://github.com/user-attachments/assets/c5dc8cfe-f0c1-4e13-a843-0e017b595d65)

**Reply by vsonnier** [reply]
This one is easy :  vkQuale supports FTE particules,  (from QSS) which Arcane Dimensions levels implement. Ironwail does'nt. (Yet)

**Comment by ghost** [comment]
Here's another example.

![Arcane Dimensions 2 vkQuake Ironwail Comparison](https://github.com/user-attachments/assets/9ce613cd-a96e-41b4-8887-c0c9ccc0a983)

**Comment by ghost** [comment]
Thanks for the feedback! Unfortunate that development for vkQuake has been stopped. I look forward to the continued development of Ironwail! I really like the improvements in the UI, particularly the mod download feature. That is a really nice touch!

**Reply by vsonnier** [reply]
> Unfortunate that development for vkQuake has been stopped. 
[Are you sure ?](https://github.com/Novum/vkQuake/releases/tag/1.31.3)

> I look forward to the continued development of Ironwail! I really like the improvements in the UI, particularly the mod download feature. That is a really nice touch!

I agree ! I picked up lots of interesting improvements from Ironwail last year, but my goal is not to make vkQuake as complete as Ironwail feature-wise (too much work anyway, not every feature has great added value IMHO). Still, the FTE particles feature is great and was the main motivation for me to continue improving vkQuake in the first place after the original author paused developpement.

**Comment by ghost** [comment]
With regard to vkQuake's development - YES, I AM COMPLETELY DISAPPOINTED! OMG vkQuake and vkQuake-RT are my absolute go to engines and have been for a long time. I use the RT version as much as I can particularly with the original Quake games! The first thing I do is try a new game with the RT version! I just love them! Certainly no offense to the IronWail team, but something about the visuals in vkQuake that just takes the lead for me. Something about the vulkan API??? I don't know - I don't have that level of technical expertise to be able to articulate it intelligently. I just see what I like - it brings a smile to my face!
